sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,263 @@
+
+        /// <summary>
+        ///   It cleans the resources that are hold by the <see cref="TransportProducerPool" />.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task CloseAsync()
+        {
+            var pendingCloses = new List<Task>();
+
+            foreach (var producer in Pool.Values)
+            {
+                pendingCloses.Add(producer.Producer.CloseAsync(CancellationToken.None));
+            }
+
+            Pool.Clear();
+            ExpirationTimer.Dispose();
+
+            await Task.WhenAll(pendingCloses).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        ///   It returns a <see cref="TimerCallback" /> that will search for all the expired <see cref="PoolItem"/>
+        ///   in the <see cref="Pool"/> and will try to close those that have expired.
+        /// </summary>
+        ///
+        internal TimerCallback PerformExpiration()
+        {
+            return _ =>
+            {
+                // Capture the timestamp to use a consistent value.
+                var now = DateTimeOffset.UtcNow;
+
+                foreach (var key in Pool.Keys)
+                {
+                    if (Pool.TryGetValue(key, out var poolItem))
+                    {
+                        if (poolItem.RemoveAfter <= now)
+                        {
+                            if (Pool.TryRemove(key, out var _) && !poolItem.ActiveInstances.Any())
+                            {
+                                // At this point the pool item may have been closed already
+                                // if there was a context switch between the if conditions
+                                // and the pool item clean up kicked in
+
+                                poolItem.Producer.CloseAsync(CancellationToken.None).GetAwaiter().GetResult();
+                            }
+                        }
+                    }
+                }
+            };
+        }
+
+        /// <summary>
+        ///   An item of the pool, adding tracking information to a <see cref="TransportProducer"/>
+        /// </summary>
+        ///
+        internal class PoolItem
+        {
+            /// <summary>The period of inactivity after which an item would be removed from the pool</summary>
+            public static readonly TimeSpan DefaultRemoveAfterDuration = TimeSpan.FromMinutes(10);
+
+            /// <summary>
+            ///   A <see cref="TransportProducer"/> assigned to a partition.
+            /// </summary>
+            ///
+            public TransportProducer Producer { get; }
+
+            /// <summary>
+            ///   A set of unique identifiers used to track which instances of a <see cref="PoolItem"/> are active.
+            /// </summary>
+            ///
+            public ConcurrentDictionary<string, byte> ActiveInstances { get; } = new ConcurrentDictionary<string, byte>();
```

Comment:
It seems we are not using the byte `value` of the dictionary entries. Did you consider using a `ConcurrentBag` instead?

Reply:
I have to say that most of the credit for this solution should go to @jsquire and I just helped really. We came out with similar designs but it was his merit to push my naive solution into a lock-free solution.

That said, @jsquire would you be so kind as to give me some help here? 😄 

When I looked at the usage of ```ConcurrentDictionary``` I thought you chose it on the basis of it fitting best with the lock-free design principle that we are trying to enforce to our solution?

However, ```ConcurrentBag``` would keep separate storage for each thread. I wonder if you think it would be more often the case that a different thread would invoke ```TransportProducerPool::GetPartitionProducer``` (call to ```TryAdd```) from the one that would invoke ```PooledProducer::CleanUp``` (call to ```TryRemove```)?